staging/src/k8s.io/apimachinery/pkg/util/cache/expiring.go

```diff
@@ -0,0 +1,192 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"container/heap"
+	"context"
+	"sync"
+	"time"
+
+	utilclock "k8s.io/apimachinery/pkg/util/clock"
+)
+
+// NewExpiring returns an initialized expiring cache that's ready to use. The
+// GC thread will shutdown when the passed context is closed.
+func NewExpiring(ctx context.Context) *Expiring {
+	return NewExpiringWithClock(ctx, utilclock.RealClock{})
+}
+
+// NewExpiringWithClock is like NewExpiring but allows passing in a custom
+// clock for testing.
+func NewExpiringWithClock(ctx context.Context, clock utilclock.Clock) *Expiring {
+	e := &Expiring{
+		clock: clock,
+		m:     make(map[interface{}]entry),
+	}
+	go e.start(ctx)
```

Comment:
Maybe it'd be better to let the constructor deal with starting the goroutine?

That way they could e.g. wait for it to stop after they close the channel.

It's also slightly weird to pass a context into a contstructor since it makes it look like it's going to store it.